c10/core/impl/InlineEvent.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include "c10/core/impl/EventInterface.h"
+#include "c10/core/DeviceType.h"
+#include "c10/core/Stream.h"
+#include "c10/util/Exception.h"
+
+namespace c10 {
+namespace impl {
+
+template <typename T>
+struct InlineEvent final : public EventInterface {
+
+  InlineEvent() = delete;
+  InlineEvent(
+    const DeviceType _device_type,
+    const EventFlag _flag = EventFlag::PYTORCH_DEFAULT)
+  : backend_{_device_type},
+    device_type_{_device_type},
+    flag_{_flag} { }
+
+  // Copy constructor and copy assignment operator (deleted)
+  InlineEvent(const InlineEvent&) = delete;
+  InlineEvent& operator=(const InlineEvent&) = delete;
+
+  // Move constructor and move assignment operator
+  InlineEvent(InlineEvent&& other)
+  : InlineEvent(other.device_type_, other.flag_) {
+    swap(std::move(other));
+  }
+  InlineEvent& operator=(InlineEvent&& other) {
+    swap(std::move(other));
+    return *this;
+  }
+
+  void swap(InlineEvent&& other) {
+    std::swap(event_, other.event_);
+    std::swap(backend_, other.backend_);
+    std::swap(device_type_, other.device_type_);
+    std::swap(device_index_, other.device_index_);
+    std::swap(flag_, other.flag_);
+    std::swap(was_marked_for_recording_, other.was_marked_for_recording_);
+  }
+
+  ~InlineEvent() noexcept {
+    if (event_) backend_.destroyEvent(event_, device_index_);
+  }
+
+  DeviceType device_type() const noexcept { return device_type_; }
+  DeviceIndex device_index() const noexcept { return device_index_; }
+  EventFlag flag() const noexcept { return flag_; }
+  bool was_marked_for_recording() const noexcept { return was_marked_for_recording_; }
+
+
+  void recordOnce(const Stream& stream) {
+    if (!was_marked_for_recording_) record(stream);
+  }
+
+  void record(const Stream& stream) {
+    TORCH_CHECK(
+      stream.device_type() == device_type_,
+      "Event device type ",
+      DeviceTypeName(device_type_),
+      " does not match recording stream's device type ",
+      DeviceTypeName(stream.device_type()),
+      ".");
+
+    backend_.record(&event_, stream, device_index_, flag_);
+    was_marked_for_recording_ = true;
+    device_index_ = stream.device_index();
+  }
+
+  void block(const Stream& stream) const {
+    if (!was_marked_for_recording_) return;
+
+    TORCH_CHECK(
+      stream.device_type() == device_type_,
+      "Event device type ",
+      DeviceTypeName(device_type_),
+      " does not match blocking stream's device type ",
+      DeviceTypeName(stream.device_type()),
+      ".");
+
+    backend_.block(event_, stream);
+  }
+
+  bool query() const {
+    if (!was_marked_for_recording_) return true;
+    return backend_.queryEvent(event_);
+  }
+
+private:
+  void* event_ = nullptr;
```

Comment:
I guess, the reason making this a `unique_ptr` sucked was because you don't statically know the type of the `cudaEvent_t` you're pointing to.

Ignoring the question of unique_ptr versus raw pointer, it might be good to pass it in as a template argument in any case. One mistake I made with InlineDeviceGuard was that it doesn't accept CUDAStream as a template argument; which is why I had to copy paste code in `aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h` instead of just saying `using OptionalHIPGuardMasqueradingAsCUDA = c10::impl::InlineOptionalDeviceGuard<HIPGuardImplMasqueradingAsCUDA>`. The extra types help you give refined signatures in inline guard which give you the option of not making a wrapper struct. (I still like writing the wrapper struct boilerplate explicitly because it helps people understand what API they can actually use.) But we don't have to handle that in this PR.

Reply:
And it doesn't really matter here because you haven't added a CUDAEvent class in this PR.

Reply:
That's right, we don't know the type of the event. It could be a cudaEvent_t, a hipEvent_t, or some as yet unknown future event type.

Let me think about your idea of using templating to improve on a void*.

Reply:
Unfortunately, I don't think there's much we can do to remove the void* here. The DeviceGuardImplInterface would need to be templatized, too, or the event functions would need to be removed from that interface. VirtualGuardImpl relies on it, however, and the entire point of VirtualGuardImpl is that it can select a backed at runtime.